test/csuite/wt7989_compact_checkpoint/main.c

```diff
@@ -0,0 +1,311 @@
+    WT_CONNECTION *conn;
+    WT_SESSION *session;
+    pthread_t thread_checkpoint, thread_compact;
+    uint64_t file_sz_after, file_sz_before;
+
+    testutil_make_work_dir(home);
+    testutil_check(wiredtiger_open(home, NULL, conn_config, &conn));
+
+    if (stress_test) {
+        /*
+         * Set WT_TIMING_STRESS_CHECKPOINT_SLOW flag. It adds 10 seconds sleep before each
+         * checkpoint.
+         */
+        set_timing_stress_checkpoint(conn);
+    }
+
+    testutil_check(conn->open_session(conn, NULL, NULL, &session));
+
+    /* Create and populate table. Checkpoint the data after that. */
+    testutil_check(session->create(session, uri, table_config));
+
+    populate(session, uri);
+    testutil_check(session->checkpoint(session, NULL));
+
+    /*
+     * Remove 1/3 of data from the middle of the key range to let compact relocate blocks from the
+     * end of the file.
+     */
+    remove_records(session, uri);
+
+    file_sz_before = get_file_size(session, uri);
+
+    td.conn = conn;
+    td.uri = uri;
+    td.cond = NULL;
+
+    /* Spawn checkpoint and compact threads. Order is important! */
+    if (stress_test) {
+        testutil_check(pthread_create(&thread_compact, NULL, thread_func_compact, &td));
+        testutil_check(pthread_create(&thread_checkpoint, NULL, thread_func_checkpoint, &td));
+    } else {
+        /* Create and initialize conditional variable. */
+        testutil_check(__wt_cond_alloc((WT_SESSION_IMPL *)session, "compact operation", &td.cond));
+
+        testutil_check(pthread_create(&thread_checkpoint, NULL, thread_func_checkpoint, &td));
+        testutil_check(pthread_create(&thread_compact, NULL, thread_func_compact, &td));
+    }
+
+    /* Wait for the threads to finish the work. */
+    (void)pthread_join(thread_checkpoint, NULL);
+    (void)pthread_join(thread_compact, NULL);
+
+    file_sz_after = get_file_size(session, uri);
+
+    /* Cleanup */
+    if (!stress_test) {
+        __wt_cond_destroy((WT_SESSION_IMPL *)session, &td.cond);
+        td.cond = NULL;
+    }
+
+    testutil_check(session->close(session, NULL));
+
+    /* Check if there's at least 10% compaction. */
+    printf(" - Compressed file size MB: %f\n - Original file size MB: %f\n",
+      file_sz_after / (1024.0 * 1024), file_sz_before / (1024.0 * 1024));
+
+    /*
+     * FIXME-WT-8055 At the moment the assert below is commented out to prevent evergreen from going
+     * red. Please enable the assert as soon as the underlying defect is fixed and compact does its
+     * job well.
+     */
+    /*testutil_assert(file_sz_before * 0.9 > file_sz_after);*/
+}
+
+static void *
+thread_func_compact(void *arg)
+{
+    struct thread_data *td;
+    WT_SESSION *session;
+
+    td = (struct thread_data *)arg;
+
+    testutil_check(td->conn->open_session(td->conn, NULL, NULL, &session));
+
+    if (td->cond != NULL) {
+        /* Make sure checkpoint thread is initialized and waiting for the signal. Sleep for one
```

Comment:
Multi-line comment not formatted correctly.

Reply:
Done